### PR TITLE
Define a new Jenkins build script to deploy AWS resources

### PR DESCRIPTION
**Pull request by junbo75**
Signed-off-by: junbo <junbo@amazon.com>

- Added a script to deploy all the CDK applications
- defined a new tag (nightly-aws-integration-deployment) for the new Jenkins job
- Update the build configuration to define a new Jenkins job and make sure that it's executed before the periodic test job
- Verified that the job can be executed on Jenkins:
https://jenkins-pipeline.agscollab.com/job/O3DE-LY-FORK/job/LYN-5119/31/ (aws_integration_deployment job)

### CODE REVIEW

**Comment by junbo75** [comment]
new_pipeline
Attached screenshot for the new pipeline.

File: scripts/build/Platform/Windows/build_config.json
Comment (onecent1101) [inline scripts/build/Platform/Windows/build_config.json:58]: I think we may need to lock cdk commit here as well? instead of HEAD?

File: scripts/build/Platform/Windows/build_config.json
Comment (onecent1101) [inline scripts/build/Platform/Windows/build_config.json:46]: this step includes all automation test? is there a way to run aws specific automation test only?

File: scripts/build/Platform/Windows/build_config.json
Comment (onecent1101) [inline scripts/build/Platform/Windows/build_config.json:39]: this name is too generic, should we name it with more information? as it is for aws automation test?

**Review by onecent1101 — COMMENTED** [review]
General question: this setup looks like gamelift required automation test which ideally we can have a separate pipeline. I remember for aws automation test, we just need a separate job to deploy resources independently not a separate pipeline?
It is ok to have a separate pipeline, but we should run aws automation test only, but right now it still runs all automation tests?

File: scripts/build/Platform/Windows/build_config.json
Comment (lmbr-pip) [reply scripts/build/Platform/Windows/build_config.json:39]: Shouldn't the vs version be a parameter that comes in? If we get VS 2021 (which is currently in preview) would we duplicate this setup?

File: scripts/build/Platform/Windows/deploy_cdk_applications.cmd
Comment (lmbr-pip) [inline scripts/build/Platform/Windows/deploy_cdk_applications.cmd:16]: Why are we not doing this inside a venv? We are impacting the machine running this?

File: scripts/build/Platform/Windows/build_config.json
Comment (junbo75) [reply scripts/build/Platform/Windows/build_config.json:58]: This is a build parameter people can change. By default, we want to use the latest unless someone specify a commit ID.

File: scripts/build/Platform/Windows/deploy_cdk_applications.cmd
Comment (lmbr-pip) [inline scripts/build/Platform/Windows/deploy_cdk_applications.cmd:17]: nit: Code is assuming a) Node is installed and b) Its a range of valid of NodeJS versions. This should be called out in comments

File: scripts/build/Platform/Windows/build_config.json
Comment (junbo75) [reply scripts/build/Platform/Windows/build_config.json:46]: What I'm doing is to remove periodic_test_profile_vs2019 from the nightly build but add this new pipeline instead. We want to make sure that the new job is executed before the periodic_test to make sure that AWS resources are deployed ahead. In this way, we don't need to separate AWS tests and put them in a new job/pipeline.

File: scripts/build/Platform/Windows/build_config.json
Comment (junbo75) [reply scripts/build/Platform/Windows/build_config.json:39]: This is not for aws automation test only. Define this pipeline to make sure that aws resources are deployed before we run all the periodic tests. I removed the "nightly-incremental" and "nightly-clean" tags on the "periodic_test_profile_vs2019" job to avoid it running on the nightly build since it needs to be replaced by the "periodic_test_profile_vs2019_pipe" job.

**Review by lmbr-pip — COMMENTED** [review]
LG.

PR should call out this is Windows only, though I am not sure why we couldn't do this on all platforms.

Consider adding notes to the periodic runbook about debugging this new step.

**Comment by junbo75** [comment]
> General question: this setup looks like gamelift required automation test which ideally we can have a separate pipeline. I remember for aws automation test, we just need a separate job to deploy resources independently not a separate pipeline?
> It is ok to have a separate pipeline, but we should run aws automation test only, but right now it still runs all automation tests?

We want a separate job to deploy resources but need to make sure that it's executed before the periodic tests to make sure that AWS resources are ready before the periodic tests start. The new pipeline will replace the existing "periodic_test_profile_vs2019" job on the nightly build.

For GameLift, you can build the dedicated server and upload to GameLift inside the new "aws_integration_deployment" job. If the client you use is for the MultiplePlayerSample project instead of AutomatedTesting, you may need to define another job for GameLift test only and create a separate pipeline to run the tests.

File: scripts/build/Platform/Windows/build_config.json
Comment (junbo75) [reply scripts/build/Platform/Windows/build_config.json:39]: Seems like the current build configuration are all hardcoding the vs version. You can copy this block and change the corresponding vs version to define a new job. I agree that we should consider to make vs version a parameter for all the Jenkins job here.

File: scripts/build/Platform/Windows/deploy_cdk_applications.cmd
Comment (junbo75) [reply scripts/build/Platform/Windows/deploy_cdk_applications.cmd:16]: We can do this inside a venv. We're using the O3DE python for our current automation tests and I just wanted to keep consistency on that.

File: scripts/build/Platform/Windows/deploy_cdk_applications.cmd
Comment (junbo75) [reply scripts/build/Platform/Windows/deploy_cdk_applications.cmd:17]: Sure. I'll comment on the code.

**Comment by junbo75** [comment]
> LG.
> 
> PR should call out this is Windows only, though I am not sure why we couldn't do this on all platforms.
> 
> Consider adding notes to the periodic runbook about debugging this new step.

All our current CDK test fixtures are defined for Windows only (AutomatedTesting\Gem\PythonTests\AWS\Windows\cdk) and we haven't tested the CDK work on other platforms. Also I'm not quite sure whether the current infra setup supports defining a common job for all the platforms since it seems that all the jobs are platform specific (o3de\scripts\build\Platform).

All the code are under the o3de\scripts\build\Platform\Windows folder and the build file is a Windows batch file, which already shows that the code are for Windows only. I can add a note in the batch file as well.

I'll update the runbook once I get approvals from infra for this change.

File: scripts/build/Platform/Windows/build_config.json
Comment (junbo75) [reply scripts/build/Platform/Windows/build_config.json]: Removed the pipeline and add a new tag so that the resource deployment can be run on a separate Jenkins job before nightly build.

File: scripts/build/Platform/Windows/deploy_cdk_applications.cmd
Comment (spham-amzn) [reply scripts/build/Platform/Windows/deploy_cdk_applications.cmd:24]: A straight call to python will invoke the build machine's python. If you want to run O3DE's python, you will need to reference the python\python.cmd wrapper script. But since you are creating a venv here anyways, I guess its safe

File: scripts/build/Platform/Windows/deploy_cdk_applications.cmd
Comment (spham-amzn) [inline scripts/build/Platform/Windows/deploy_cdk_applications.cmd:48]: I would add ERRORLEVEL checks to these calls as well, since any one of them can fail

File: scripts/build/Platform/Windows/deploy_cdk_applications.cmd
Comment (spham-amzn) [inline scripts/build/Platform/Windows/deploy_cdk_applications.cmd:58]: Are multiple calls to git checkout necessary? Can't this be done at the %GEM_DIRECTORY% root?

File: scripts/build/Platform/Windows/deploy_cdk_applications.cmd
Comment (junbo75) [reply scripts/build/Platform/Windows/deploy_cdk_applications.cmd:24]: I've added the o3de python path to the environment variable using:
"SET PATH=%SOURCE_DIRECTORY%\python;%PATH%"

Also created a venv here with the latest change.

File: scripts/build/Platform/Windows/deploy_cdk_applications.cmd
Comment (junbo75) [reply scripts/build/Platform/Windows/deploy_cdk_applications.cmd:58]: I'm trying to checkout the CDK code and leave all the other gem code as it is. We may need to build dedicated server with this job later when the automation tests for the GameLift gem are added. Checking out the entire folder may revert the client side C++ code to a previous version as well, which seems unexpected.

File: scripts/build/Platform/Windows/deploy_cdk_applications.cmd
Comment (junbo75) [reply scripts/build/Platform/Windows/deploy_cdk_applications.cmd:48]: Will add the checks. Thanks.

File: scripts/build/Platform/Windows/build_config.json
Comment (shirangj) [inline scripts/build/Platform/Windows/build_config.json:42]: Does this need to run in weekly-build-metrics?

Comment (junbo75) [reply scripts/build/Platform/Windows/build_config.json:42]: I thought the tag was for collecting metrics instead of an actual Jenkins job. Will remove it then. Thanks.

**Review by spham-amzn — APPROVED** [review]

**Review by shirangj — APPROVED** [review]

File: scripts/build/Platform/Windows/build_config.json
Comment (onecent1101) [reply scripts/build/Platform/Windows/build_config.json:48]: Minor: I think we want to lock it to specific commit for cdk deployment which we have discussed in the infra team meeting. (we should pick a commit id)

**Review by onecent1101 — APPROVED** [review]

File: scripts/build/Platform/Windows/build_config.json
Comment (junbo75) [reply scripts/build/Platform/Windows/build_config.json:48]: We can address this in a future PR once the Jenkins job is set up. We will find a stable version (currently the latest should be stable) and lock it.

File: scripts/build/Platform/Windows/build_config.json
Comment (amzn-changml) [inline scripts/build/Platform/Windows/build_config.json:47]: It might be better to parameterize this ARN/Region and move it to a Jenkins envvar. If you move between production and sandbox Jenkins, you want some flexibility to have different environment account and roles.

You can refer to envvars with bangs like this
```
      "O3DE_AWS_DEPLOY_REGION": "!AWSI_DEPLOY_REGION!",
      "ASSUME_ROLE_ARN": "!AWSI_DEPLOY_ROLE_ARN!",
```

I can to work with you on the naming convention if you want to do this.

Comment (junbo75) [reply scripts/build/Platform/Windows/build_config.json:47]: Sure. Will sync offline to set the envvar.

Comment (brianherrera) [reply scripts/build/Platform/Windows/build_config.json:47]: +1 We should avoid hard-coding ARNs for specific AWS resources. This will be different depending if we run this in our fork pipeline or in the O3DE pipeline which are running in separate AWS accounts. This also prevents maintainers that fork O3DE from running these tests in their account.

Comment (shirangj) [reply scripts/build/Platform/Windows/build_config.json:47]: To move these parameters to Jenkins envvar, it might be better to create a new Jenkinsfile instead of adding them to the Jenkinsfile used by AR builds.

Comment (junbo75) [reply scripts/build/Platform/Windows/build_config.json:47]: Just updated the PR to use envvar. Currently, this job just deploys the CDK applications and will not test the engine. However, the CDK application code are in the main repo and we don't have a separate awsi repo for them. Also we talked about the use case for GameLift during the office hour. After we add the automation tests for GameLift gem, we do want to build the dedicated server and upload it with this new Jenkins job. In that case, it needs the engine code.

File: scripts/build/Platform/Windows/pipeline.json
Comment (shirangj) [inline scripts/build/Platform/Windows/pipeline.json:8]: The parameters defined in pipeline.json won't show up as a Jenkins build parameters, so we still have the same problem with putting them in build_config.json.

To make these parameter be Jenkins build parameters, they need to be defined in Jenkinsfile, but by doing this, all AR jobs will have these Jenkins build parameters because all AR jobs are using the same Jenkinsfile. I'm not against this approach but it might cause some confusion to people who manually triggers an AR build.

Comment (junbo75) [reply scripts/build/Platform/Windows/pipeline.json]: Synced offline. We will create a separate jenkinsfile pipeline instead.

File: scripts/build/Platform/Windows/build_config.json
Comment (junbo75) [reply scripts/build/Platform/Windows/build_config.json]: Talked with Mike and Shirang offline. We will create a separate jenkinsfile pipeline for the resource deployment instead.

**Pull request merged**
